h/tasks/es_sync.py

```diff
@@ -0,0 +1,107 @@
+import os
+
+from dateutil.parser import isoparse
+from psycopg2.errors import LockNotAvailable
+
+from h.celery import celery, get_task_logger
+from h.models import Annotation, EsSyncTask
+from h.search.index import BatchIndexer
+
+log = get_task_logger(__name__)
+
+
+@celery.task
+def es_sync():
+    """Synchronize annotations from Postgres to Elasticsearch."""
+
+    request = celery.request
+    db = celery.request.db
+    es = celery.request.es
+
+    annotation_ids_to_index = set()
+    completed_messages = set()
+
+    messages = get_messages_from_queue(db)
+
+    if not messages:
+        log.info("No annotations to sync")
+        return
+
+    annotations_from_db = get_annotations_from_db(db, messages)
+    annotations_from_es = get_annotations_from_es(es, annotations_from_db)
+
+    invalid_messages = [
+        message
+        for message in messages
+        if message.annotation_id
+        not in (annotation.id for annotation in annotations_from_db)
+    ]
+    if invalid_messages:
+        log.info(
+            f"Deleting messages from the queue because their annotations have been deleted from the DB: {invalid_messages}"
+        )
+        for message in invalid_messages:
+            db.delete(message)
+
+    for annotation in annotations_from_db:
+        annotation_from_es = annotations_from_es.get(annotation.id)
+        matching_messages = [
+            message for message in messages if message.annotation_id == annotation.id
+        ]
+
+        if not annotation_from_es or annotation.updated != annotation_from_es:
+            annotation_ids_to_index.add(annotation.id)
+            for message in matching_messages:
+                message.attempts += 1
+                if message.attempts > 2:
+                    log.info(
+                        f"Deleting a message from the queue because it failed to sync to the search index too many times: {message}"
+                    )
+                    db.delete(message)
+        else:
+            completed_messages.update(messages)
+
+    if completed_messages:
+        log.info(f"Completed messages from queue: {completed_messages}")
+        for message in completed_messages:
+            db.delete(message)
+
+    if annotation_ids_to_index:
+        log.info(f"Indexing annotations: {annotation_ids_to_index}")
+        BatchIndexer(db, es, request).index(annotation_ids_to_index)
+
+
+def get_messages_from_queue(db):
+    """Return the next block of messages to be handled, from the queue."""
+    try:
+        return (
+            db.query(EsSyncTask)
+            .order_by(EsSyncTask.id)
+            .limit(os.environ.get("ES_SYNC_TASK_LIMIT", 1000))
+            .with_for_update(nowait=True)
+            .all()
+        )
```

Comment:
This query also needs to filter out very recent messages (say added to the queue in the last 30 secs?) so that the periodic task doesn't fight with web requests. That'll require adding a `time_added` column to the queue